Add statefulness to the absolute price agent

aasseman opened the pull request:
Refer to individual commit messages for details.

aasseman committed:
feat: Add statefulness to the absolute price agent

The absolute pricing agent (price multiplier) will save its mean and
stddev state to the Postgres database, and try to restore it on
agent startup. The save state is expired after 24 hours.

Signed-off-by: Alexis Asseman <alexis@semiotic.ai>

aasseman committed:
fix: share a single PG connection pool everywhere

It's the proper way to do it.
Fixes collisions when there is simultaneous DB write operations in the price
save state.

Signed-off-by: Alexis Asseman <alexis@semiotic.ai>

aasseman committed:
chore: update asyncpg to v0.27.0

Signed-off-by: Alexis Asseman <alexis@semiotic.ai>

tumayt reviewed (commented):
I  think agents should have API calls to init schema, restore and save for their state. As autoagora should be agnostic to the internals of agents. Also SqlAlchemy may be useful for clean DDL and queries.
If this is a temporary solution till agents v2, LGTM.

aasseman commented:
> I think agents should have API calls to init schema, restore and save for their state. As autoagora should be agnostic to the internals of agents.

I agree, but I wanted a solution now and not get bogged down in the complexity of setting up the proper thing :neutral_face:

aasseman committed:
fix: update agents to fix stability issues

Updated autoagora-agents to v1.2.1 to fix the issue.
Also added a test to make sure the issue is indeed fixed.

Signed-off-by: Alexis Asseman <alexis@semiotic.ai>

aasseman commented:
Had to wait for https://github.com/semiotic-ai/autoagora-agents/releases/tag/v1.2.1 to fix some issues.
Also, rebased onto latest master.

tumaysem commented on autoagora/main.py:
```diff
@@ -35,6 +36,21 @@ async def allocated_subgraph_watcher():
 
+    try:
+        pgpool = await asyncpg.create_pool(
+            host=args.postgres_host,
```
from my previous experience, pool size should be controlled (from args ?)
https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg-api-pool

aasseman replied:
Yes, you're right, I should make this a user setting.
In general, what do you think of limiting to a single connection by default? None of the code paths are extremely time-sensitive, and I don't expect cyclical dependencies.

aasseman replied:
https://github.com/semiotic-ai/autoagora/pull/54/commits/096087fbc4ec01a763737d6b4c2714a6da12d816

aasseman committed:
feat: add user arg for # PG connections

AA maintains its own PG connections pool. It is now set at a minimum of 1
connection, and a default maximum of 1 that can be overriden through
the `--postgres-max-connections` argument.

Signed-off-by: Alexis Asseman <alexis@semiotic.ai>

tumayt reviewed (approved):
LGTM
